src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,445 @@
+{
+    uint8_t newMode = commandData.newMode;
+
+    Commands::ChangeToModeResponse::Type response;
+
+    if (!IsSupportedMode(newMode))
+    {
+        ChipLogError(Zcl, "ModeBase: Failed to find the option with mode %u", newMode);
+        response.status = static_cast<uint8_t>(StatusCode::kUnsupportedMode);
+        ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+        return;
+    }
+
+    HandleChangeToMode(newMode, response);
+
+    if (response.status == static_cast<uint8_t>(StatusCode::kSuccess))
+    {
+        UpdateCurrentMode(newMode);
+        ChipLogProgress(Zcl, "ModeBase: handleChangeToMode changed to mode %u", newMode);
+    }
+
+    ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+}
+
+// This function is called by the interaction model engine when a command destined for this instance is received.
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case ModeBase::Commands::ChangeToMode::Id:
+        ChipLogDetail(Zcl, "ModeBase: Entering handling ChangeToModeWithStatus");
+
+        HandleCommand<Commands::ChangeToMode::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { handleChangeToMode(ctx, commandData); });
+    }
+}
+
+// List the commands supported by this instance.
+CHIP_ERROR Instance::EnumerateAcceptedCommands(const ConcreteClusterPath & cluster,
+                                               CommandHandlerInterface::CommandIdCallback callback, void * context)
+{
+    callback(ModeBase::Commands::ChangeToMode::Id, context);
+    callback(ModeBase::Commands::ChangeToModeResponse::Id, context);
+    return CHIP_NO_ERROR;
+}
+
+// Implements the read functionality for complex attributes.
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Attributes::CurrentMode::Id:
+        ReturnErrorOnFailure(aEncoder.Encode(mCurrentMode));
+        break;
+    case Attributes::StartUpMode::Id:
+        if (mStartUpMode.IsNull())
+        {
+            ReturnErrorOnFailure(aEncoder.EncodeNull());
+        }
+        else
+        {
+            ReturnErrorOnFailure(aEncoder.Encode(mStartUpMode));
+        }
+        break;
+    case Attributes::OnMode::Id:
+        if (mOnMode.IsNull())
+        {
+            ReturnErrorOnFailure(aEncoder.EncodeNull());
+        }
+        else
+        {
+            ReturnErrorOnFailure(aEncoder.Encode(mOnMode));
+        }
+        break;
+    case Attributes::SupportedModes::Id:
+        if (NumberOfModes() == 0)
+        {
+            aEncoder.EncodeEmptyList();
+            return CHIP_NO_ERROR;
+        }
+
+        Instance * d   = this;
+        CHIP_ERROR err = aEncoder.EncodeList([d](const auto & encoder) -> CHIP_ERROR {
```

Comment:
As previously requested, if this can be moved to the anonymous namespace as a helper function, it would make the dispatch switch case easier to read.

Reply:
Refactored [here](https://github.com/project-chip/connectedhomeip/pull/27504/commits/36ef75bb4533874aa56f98fe85d9054284501265).